src/Sparrow/Json/Parsing/JsonParserState.cs

```diff
@@ -217,6 +219,199 @@ public void Reset()
             CurrentTokenType = JsonParserToken.None;
             Continuation = JsonParserTokenContinuation.None;
             EscapePositions.Clear();
+
+            // We will change this value to MinValue because the 0 magnitude number is valid as the content of a vector.
+            _maxLongMagnitude = long.MinValue;
+            _maxDoubleMagnitude = double.MinValue;
+        }
+
+        internal FastList<(JsonParserToken, long)> _bufferedSequence;
+        private long _maxLongMagnitude;
+        private double _maxDoubleMagnitude;
+        private bool _hasNegativeValue;
+
+        internal bool IsBufferedFloat { get; private set; }
+
+        internal void ClearBuffered()
+        {
+            _bufferedSequence.WeakClear();
+
+            // We will change this value to MinValue because the 0 magnitude number is valid as the content of a vector.
+            _maxLongMagnitude = long.MinValue;
+            _maxDoubleMagnitude = double.MinValue;
+        }
+
+        internal BlittableVectorType GetBufferedOptimalType()
+        {
+            if (IsBufferedFloat)
+            {
+                // We need to check both because we can convert longs to doubles, but not the other
+                // way around. 
+                double maxValue = Math.Max(_maxDoubleMagnitude, _maxLongMagnitude);
+                
+                // If handling floating-point values, use a simple check first.
+                return Math.Abs(maxValue) switch
+                {
+                    <= float.MaxValue => BlittableVectorType.Float,
+                    _ => BlittableVectorType.Double,
+                };
+            }
+
+            // At this point we know we are only working with integers.
+            return (_maxBufferedLongMagnitude: _maxLongMagnitude, _hasNegativeLongValue: _hasNegativeValue) switch
+            {
+                ( <= sbyte.MaxValue, true) => BlittableVectorType.SByte,
+                ( <= byte.MaxValue, false) => BlittableVectorType.Byte,
+                ( <= short.MaxValue, true) => BlittableVectorType.Int16,
+                ( <= ushort.MaxValue, false) => BlittableVectorType.UInt16,
+                ( <= int.MaxValue, true) => BlittableVectorType.Int32,
+                ( <= uint.MaxValue, false) => BlittableVectorType.UInt32,
+                (_, true) => BlittableVectorType.Int64,
+                (_, false) => BlittableVectorType.UInt64, // Fallback for unsigned long
```

Comment:
Why do we care that about `ulong` vs `long` ?

Reply:
Because we bake that in into the format we advertise. It's not the same if it is signed or unsigned for the reader.